### DISCUSSION TITLE
mqtt acl:download requiring RSA key

### DISCUSSION THREAD

**Discussion by irvintim** [uhppoted/uhppoted#17] [Q&A] [answered]
Hi:

So, my project has been moving along slowly, the next thing I want to accomplish is to be able to load up the ACLs (cards) via a Lambda job.  I'm storing the card data in DynamoDB, and then when that is updated I trigger a Lambda to grab the latest list of cards, generate a new *.acl file and tar that up and upload it to S3.  Then send an "acl:download" MQTT message with the URL for this S3 file.  Everything is working great until I get the following:

```
{
  "message": {
    "error": {
      "client-id": "test",
      "error": {
        "code": 400,
        "debug": ": no RSA public key",
        "message": "Error downloading ACL"
      },
      "method": "acl:download",
      "nonce": 6,
      "request-id": "AH1635G3",
      "server-id": "uhppoted"
    }
  },
  "hmac": "xxxx"
}
```

I found the instructions in the repo on how to create the RSA signed acl files, but I'd rather not have to do that (it's already hard enough to create a gziped tar file in a python-based Lambda job without writing anything to disk).  Is there a way to turn this "feature" off?

I am running [uhppoted-mqtt v0.8.1-beta](https://github.com/uhppoted/uhppoted-mqtt/releases/tag/v0.8.1)

Tim

**Comment by uhppoted** [comment]
Hi Tim,

Interesting system you're building there!!

I've added `mqtt.acl.verify` to the configuration so if you pull and build the master branch you should be able to disable the ACL signature verification by adding this line to _uhppoted.conf_:
```
mqtt.acl.verify = false
```

Most of the code to support it was already there - it just had to be added to the configuration. I have tested it, but only briefly so let me know if you run into anything.

re. _tar.gz_. I did **not** think anybody would ever even consider doing that in a lambda :-) ! The _tar.gz_ requirement was mostly there so that the signature was bundled with the ACL - with the no-verify option I can look into into adding support for non-archive ACL files. But that's a little more complicated and needs some thinking about.

The tracking issue for this is [#18: Add mqtt.acl.verify option to uhppoted.conf](https://github.com/uhppoted/uhppoted-mqtt/issues/6) but I'll post updates here as well.

**Reply by irvintim** [reply]
Thanks! I'll build the new rev tomorrow and give it a shot.  I'll let you know.

Yes, building a tar.gz in python3 was interesting.  Here is the resulting code if you are interested in some light reading:

```
import tarfile
import boto3
import io
import time

def make_tarfile(tarobj, files):
    def make_tarinfo(filename, filebody):
        file_tarinfo = tarfile.TarInfo(filename)
        file_tarinfo.size = len(filebody.encode('utf-8'))
        file_tarinfo.mtime = time.time()
        return(file_tarinfo)

    def make_fileobj(fileobj, filename):
        fileobj.name = filename
        fileobj.seek(0)
        return(fileobj)

    with tarfile.open(fileobj=tarobj, mode="w:gz") as tar:
        for (fname, body) in files:
            file_tarinfo = make_tarinfo(fname, body)
            file_obj = make_fileobj(io.BytesIO(body.encode('utf-8')), fname)
            logger.debug("Adding {}, size {} to tarfile".format(fname, file_tarinfo.size))
            tar.addfile(file_tarinfo, fileobj=file_obj)
    tarobj.seek(0)
    return(tarobj)
 ```
 
 In the main lambda function it was called like this:
 ```
 def lambda_function(event, context):
    ... stuff to get records and build an ACL file in memory (outfile var) ...
    tarobj = make_tarfile(io.BytesIO(), [('uhppoted.acl', output), ('signature', '')])
    s3 = boto3.client('s3')
    response = s3.put_object(Body=tarobj, Bucket=os.environ.get("S3_BUCKET"), Key='ACL-{}.tar.gz'.format(context.aws_request_id))
    ...
```

**Reply by irvintim** [reply]
I have installed the new rev and it is working great, thanks!

**Comment by uhppoted** [comment]
Great! Thanks for letting me know! 

And, thanks for the code snippet - I'll add it to the slowly growing knowledge base. It's interestingly terse in _Python_ - the _Go_ version is at least 50% more code (admittedly some of that is error handling but still).

BTW, just a heads up - as a first step to supporting non-archive formats, the latest version in the master branch has reworked `mqtt.acl.verify` to be a string with the options `none`, `not-empty` and `RSA` (not case-sensitive). It also (temporarily) allows `false` so that your system will continue to work if you inadvertently pull the latest and forget to change it in _uhppoted.conf_.

I remember now why I decided on the _signed .tar.gz_ format .. there were just a lot fewer "_but what if ..._" things to think about :-).

**Comment by uhppoted** [comment, marked as answer]
Just to wrap this up .. _uhppoted-mqtt_ now supports downloading an ACL as an unsigned  TSV file. These are the changes that are needed:

1. _uhppoted.conf_
```
...
mqtt.acl.verify = not-empty, RSA
...
```
(you can use `none` but entirely at your own risk - it's not recommended)

2. The request message `mime-type` should be set to `text/tab-separated-values`:
```
mqtt publish --topic 'uhppoted/gateway/requests/acl/acl:download'
                      --message '{ "message": { "request": { 
                                           "url": "file://../runtime/mqttd/hogwarts.acl", 
                                           "mime-type": "text/tab-separated-values", 
                                           "client-id": "QWERTY", 
                                           "reply-to": "uhppoted/reply/97531", 
                                           "request-id": "AH173635G3" }}}'
```
(the `mime-type` defaults to `application/x-gzip` which is one of the unofficial mime-types for tar.gz files).

This update is only in the development branch at the moment - it should make it into the next release, although that is still  a couple of weeks away.

**Comment by irvintim** [comment]
Awesome, thank you.  This is a big help.

**Comment by uhppoted** [comment]
Released in [v0.8.2](https://github.com/uhppoted/uhppoted-mqtt/releases/tag/v0.8.2).

BTW, I (entirely inadvertently) left the support for `mqtt.acl.verify = false` in the codebase, but the intention is to remove it in the next release so you want to change it to officially supported `mqtt.acl.verify = none`.
